### PR TITLE
Basic facet support

### PR DESCRIPTION
**Pull request by meg-codes**
This PR provides basic facet support to querysets for Princeton-CDH/mep-django#199. There are two main methods added, `facet` to enable facets and set a list of facet fields, and `facet_fields` to return a results facet fields, and their counts, as a dictionary.

### CODE REVIEW

File: parasolr/query.py [hunk 1]
```diff
@@ -123,6 +127,22 @@ def count(self) -> int:
         query_opts['rows'] = 0
         return self.solr.query(**query_opts, wrap=False)['response']['numFound']
 
+    def facet_fields(self) -> dict:
+        """Return a dictionary of facets and their values/counts (also as a
+        dicitionary).
```
Comment (rlskoeser) [inline parasolr/query.py]: typo in second dictionary; seems like there should be a clearer way to describe this. Maybe explicitly use key/value? dictionary keyed on facet field, value is an `OrderedDict` of facet values and counts.

File: parasolr/query.py [hunk 2]
```diff
@@ -123,6 +127,22 @@ def count(self) -> int:
         query_opts['rows'] = 0
         return self.solr.query(**query_opts, wrap=False)['response']['numFound']
 
+    def facet_fields(self) -> dict:
```
Comment (rlskoeser) [inline parasolr/query.py]: I wonder if `get_facets` or `get_facet_fields` would be clearer here, to distinguish more from the method where you're applying/adding facets? We're using that for `get_highlighting`

File: parasolr/query.py [hunk 3]
```diff
@@ -123,6 +127,22 @@ def count(self) -> int:
...
+        """Return a dictionary of facets and their values/counts (also as a
+        dicitionary).
+        """
+        if self._result_cache is not None:
+            # wrap to process facets and return as dictionary
+            # for Django template support
+            qr = QueryResponse(self._result_cache)
```
Comment (rlskoeser) [inline parasolr/query.py:148]: I see you're using the `QueryResponse` class to take advantage of the facet handling, that makes sense. Do you think we can adjust so that `SolrQuerySet` can use `QueryResponse` everywhere? 

I can't remember why I wasn't using it, was it because of `AttrDict`? Any thoughts on how to get around that?

File: parasolr/query.py [hunk 4]
```diff
@@ -158,6 +178,30 @@ def filter(self, *args, **kwargs) -> 'SolrQuerySet':
 
         return qs_copy
 
+    def facet(self, *args: str, **kwargs) -> 'SolrQuerySet':
+        """
+        Return a new SolrQuerySet with Solr faceting enabled and facet.field
```
Comment (rlskoeser) [inline parasolr/query.py]: This doesn't seem as readable as I'd like (I think too technical / specific to what's going on underneath?). Maybe start with something like:
>  Request facets for the specified fields.

And then your existing description could follow.

File: parasolr/query.py [hunk 5]
```diff
@@ -158,6 +178,30 @@ def filter(self, *args, **kwargs) -> 'SolrQuerySet':
...
+        parameter set. Does not yet support ranged faceting.
```
Comment (rlskoeser) [inline parasolr/query.py]: I would drop the *yet* - suspect we'll want a different method to configure range facets

File: parasolr/query.py [hunk 6]
```diff
@@ -158,6 +178,30 @@ def filter(self, *args, **kwargs) -> 'SolrQuerySet':
...
+
+        Subsequent calls will reset the facet.field to the last set of
```
Comment (rlskoeser) [inline parasolr/query.py:202]: Suggest rewording to say it will override any previously set facets.

File: parasolr/query.py (hunk 3 above)
Comment (rlskoeser) [reply parasolr/query.py:148]: A thought, if `AttrDict` is the reason - could `QueryResponse` store both the dict and the wrapping attrdict, maybe? Then we have access to attrdict for convenience but not stuck with it.

File: parasolr/query.py [hunk 7]
```diff
@@ -158,6 +178,30 @@ def filter(self, *args, **kwargs) -> 'SolrQuerySet':
...
+            qs = qs.facet('item_type')
+
+        would result in `item_type` being the only facet field.
+        """
+        qs_copy = self._clone()
+
+        # enable faceting
+        qs_copy.facet_opts['facet'] = True
+        # add args as a list to facet.field
+        qs_copy.facet_opts['facet.field'] = list(args)
```
Comment (rlskoeser) [inline parasolr/query.py]: Isn't `args` already a list?

File: parasolr/solr/client.py [hunk 8]
```diff
@@ -19,13 +19,15 @@
 # despite not being hugely Pythonic, for consistency with Solr's responses
 # and API documentation.
 
-class QueryReponse:
+class QueryResponse:
```
Comment (rlskoeser) [inline parasolr/solr/client.py:37]: Ha, good catch. (How did we both miss that typo?)

File: parasolr/query.py [hunk 9]
```diff
@@ -123,6 +127,22 @@ def count(self) -> int:
...
 
+    def facet_fields(self) -> dict:
+        """Return a dictionary of facets and their values/counts (also as a
+        dicitionary).
+        """
+        if self._result_cache is not None:
+            # wrap to process facets and return as dictionary
+            # for Django template support
+            qr = QueryResponse(self._result_cache)
+            return dict(qr.facet_counts.facet_fields)
```
Comment (rlskoeser) [inline parasolr/query.py]: Why is this casting the return to a `dict`?

File: parasolr/query.py [hunk 10]
```diff
@@ -158,6 +178,30 @@ def filter(self, *args, **kwargs) -> 'SolrQuerySet':
...
+
+        would result in `item_type` being the only facet field.
+        """
+        qs_copy = self._clone()
+
+        # enable faceting
+        qs_copy.facet_opts['facet'] = True
+        # add args as a list to facet.field
+        qs_copy.facet_opts['facet.field'] = list(args)
+
```
Comment (rlskoeser) [inline parasolr/query.py:218]: What about other facet options? Sort, min count. Could handle the way we're doing additional highlighting options, just use `kwargs`...

I'm less certain how to handle field-specific facet options, but maybe we hold off on that until we have a specific use case.

File: parasolr/tests/test_query.py [hunk 11]
```diff
@@ -90,6 +91,60 @@ def test_count(self):
         # cache should not be populated
         assert not sqs._result_cache
 
+    def test_facet_fields(self):
+        mocksolr = Mock(spec=SolrClient)
```
Comment (rlskoeser) [inline parasolr/tests/test_query.py:110]: I'm not sure this test is actually testing the right things - it should be completely agnostic as to the specifics of the Solr response. Maybe better to patch `QueryResponse` and check that it is called or not called as expected? Check that you get the right return type, not the specific values.

File: parasolr/tests/test_query.py [hunk 12]
```diff
@@ -117,6 +172,22 @@ def test_filter(self):
...
+        faceted_qs = sqs.facet(*facet_list)
+        # faceting should be set on
+        assert faceted_qs.facet_opts['facet']
+        assert faceted_qs.facet_opts['facet.field'] == facet_list
+        # subsequents calls to facet should simply reset list
+        facet_list = ['foobars']
+        faceted_qs = faceted_qs.facet(*facet_list)
+        assert faceted_qs.facet_opts['facet']
+        assert faceted_qs.facet_opts['facet.field'] == facet_list
+
```
Comment (rlskoeser) [inline parasolr/tests/test_query.py]: Important case you're missing here - facet options should be preserved across other methods that return a new queryset, and should not modify the original. You're kind of missing it because you're testing that new facets replace previous facets. (Not sure what that should do for any previous facet options... do they reset too?)

I expect this to fail because I don't see any logic in the `_clone` method to copy facet options. Might be a good opportunity to write the failing test first.

File: parasolr/query.py (hunk 10 above)
Comment (rlskoeser) [reply parasolr/query.py:218]: If we want field-specific facets, _that_ might be a method like `facet_field` where you specify the field and then the field-specific facet options. (I don't think we need this right away, but does confirm my opinion to change the existing `facet_fields` method to `get_facets` or similar.)

File: parasolr/query.py [hunk 13]
```diff
@@ -105,6 +106,9 @@ def query_opts(self) -> Dict[str, str]:
             for key, val in self.highlight_opts.items():
                 query_opts['hl.%s' % key] = val
 
+        if self.facet_opts:
```
Comment (rlskoeser) [inline parasolr/query.py]: I don't see a test case for this addition, please add.

File: parasolr/query.py [hunk 14]
```diff
@@ -123,6 +127,22 @@ def count(self) -> int:
...
+        if self._result_cache is not None:
+            # wrap to process facets and return as dictionary
+            # for Django template support
+            qr = QueryResponse(self._result_cache)
+            return dict(qr.facet_counts.facet_fields)
+        # since we just want a dictionary of facet fields, don't populate
+        # the result cache and do a wrapped request with an explicit cast
+        # to dict, no rows needed.
+        query_opts = self.query_opts()
+        query_opts['rows'] = 0
```
Comment (rlskoeser) [inline parasolr/query.py:155]: I know you borrowed this from my code - I _think_ that kwargs passed in now take precedence over generated query opts, so it ought to be possible to just call `query` with the `rows=0` override and let it use all the other defaults.

It occurs to me the `count` query could probably drop faceting and highlighting options too...

**Review by rlskoeser — CHANGES_REQUESTED** [review]
This looks pretty good. I have several suggestions for wording and names that I think might make things clearer and more readable. The one big problem is that you need to update the `_clone` method to make sure that facet options are persisted across methods that return new queryset instances.  It looks like everything else is pretty much working, just a matter of refining.

Your changes also prompted me to notice some things from the existing code that we might be able to improve. The `QueryResponse` one is probably the biggest, and we might decide we don't want to tackle it yet, but I'm interested to hear what you think about it.

File: parasolr/query.py (hunk 2 above)
Comment (meg-codes) [reply parasolr/query.py]: I think `get_facets` makes sense here, so let's go with that.

File: parasolr/query.py (hunk 1 above)
Comment (meg-codes) [reply parasolr/query.py]: I like this, also the use of `OrderedDict` because we'll want it anyway.

File: parasolr/query.py (hunk 4 above)
Comment (meg-codes) [reply parasolr/query.py]: Done.

File: parasolr/query.py (hunk 5 above)
Comment (meg-codes) [reply parasolr/query.py]: Also done.

File: parasolr/query.py (hunk 10 above)
Comment (meg-codes) [reply parasolr/query.py:218]: I had somewhat deliberately held off on the facet kwargs because I wasn't sure if we wanted to add complexity beyond what was needed for the story that we're making these adjustments for. However, just doing an update for kwargs would allow for using them without major issues, so I'll add that.

File: parasolr/query.py (hunk 7 above)
Comment (meg-codes) [reply parasolr/query.py]: `args` is actually a tuple, and given that we might as some point find ourselves modifying it, I thought it best to cast it to a mutable sequence explicitly. I'll note why explicitly, which I'd meant to do.

File: parasolr/query.py (hunk 9 above)
Comment (meg-codes) [reply parasolr/query.py]: This is to drag it out of being an `AttrDict` (they're recursive for Dict-like collections). I'm feeling increasingly less convinced about the utility of `AttrDict` given its issues with Django templating... but it does read more easily.

File: parasolr/query.py (hunk 2 above)
Comment (rlskoeser) [reply parasolr/query.py]: I like that - and we might be able to get all different facets in a single method even if we have different methods for requesting them.

File: parasolr/query.py (hunk 7 above)
Comment (rlskoeser) [reply parasolr/query.py]: I meant to check if it was a tuple. Makes sense, but a note would be helpful.

File: parasolr/query.py [hunk 15]
```diff
@@ -261,6 +318,7 @@ def _clone(self) -> 'SolrQuerySet':
         qs_copy.start = self.start
         qs_copy.stop = self.stop
         qs_copy.highlight_field = self.highlight_field
+        #qs.copy.facet_opts = self.facet_opts
```
Comment (rlskoeser) [inline parasolr/query.py]: I see this is commented out and that worries me. It should be handled the same way that other dict and list options are. I can add a line to your tests that shows the problem if that's helpful - it appears to be working right now, but you're setting facets by reference so all versions of the queryset share them.

File: parasolr/tests/test_query.py [hunk 16]
```diff
@@ -117,6 +174,29 @@ def test_filter(self):
...
+        faceted_qs = faceted_qs.filter(foo='bar')
+        assert faceted_qs.facet_opts['facet']
+        assert faceted_qs.facet_opts['facet.field'] == facet_list
+        # subsequents calls to facet should simply reset list
+        facet_list = ['foobars']
+        faceted_qs = faceted_qs.facet(*facet_list)
+        assert faceted_qs.facet_opts['facet']
+        assert faceted_qs.facet_opts['facet.field'] == facet_list
+        # kwargs should simply be set in facet opts
+        faceted_qs = faceted_qs.facet(*facet_list, **{'facet.sort': 'count'})
```
Comment (rlskoeser) [inline parasolr/tests/test_query.py]: Ah! Seeing this now I see I was unclear. I want to be able to say something like `sqs.facet('member', sort='count')`. If you look at the highlighting you'll see it works that way - the prefix is added in `query_opts`

File: parasolr/query.py [hunk 17]
```diff
@@ -158,6 +187,34 @@ def filter(self, *args, **kwargs) -> 'SolrQuerySet':
...
+
+            qs = queryset.facet('person_type', 'age')
+            qs = qs.facet('item_type')
+
+        would result in `item_type` being the only facet field.
+        """
+        qs_copy = self._clone()
+
+        # enable faceting
+        qs_copy.facet_opts['facet'] = True
```
Comment (rlskoeser) [inline parasolr/query.py]: Can we just infer `facet=true` if `facet_opts` is not empty when we generate query opts?

File: parasolr/solr/tests/test_admin.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
+from unittest.mock import Mock, patch
```
Comment (rlskoeser) [inline parasolr/solr/tests/test_admin.py:1]: Thank you for splitting out the solr tests!

**Review by rlskoeser — APPROVED** [review]
